injector/injector.go

```diff
@@ -0,0 +1,275 @@
+				encoder.Flush()
+				if !errorTagFound {
+					ti.addErrorTrackingEvent(&outputXML)
+					errorTagFound = true
+				}
+				continue
+			case "NonLinearAds":
+				if injectTracker {
+					injectTracker = false
+					encoder.Flush()
+					ti.addTrackingEvent(&outputXML, creativeId, true)
+					if !nonLinearTagFound && wrapperTagFound {
+						ti.addNonLinearClickTrackingEvent(&outputXML, creativeId, true)
+					}
+					encoder.EncodeToken(tt)
+				}
+			case "Linear":
+				if injectVideoClicks {
+					injectVideoClicks = false
+					encoder.Flush()
+					ti.addClickTrackingEvent(&outputXML, creativeId, true)
+				}
+				if injectTracker {
+					injectTracker = false
+					encoder.Flush()
+					ti.addTrackingEvent(&outputXML, creativeId, true)
+				}
+				encoder.EncodeToken(tt)
+				continue
+			case "InLine", "Wrapper":
+				wrapperTagFound = false
+				inlineWrapperTagFound = true
+				encoder.Flush()
+				if !impressionTagFound {
+					ti.addImpressionTrackingEvent(&outputXML)
+				}
+				impressionTagFound = false
+				if !errorTagFound {
+					ti.addErrorTrackingEvent(&outputXML)
+				}
+				errorTagFound = false
+				encoder.EncodeToken(tt)
+			case "NonLinear":
+				encoder.Flush()
+				ti.addNonLinearClickTrackingEvent(&outputXML, creativeId, false)
+				nonLinearTagFound = true
+				encoder.EncodeToken(tt)
+			case "Companion":
+				companionTagFound = true
+				encoder.Flush()
+				ti.addCompanionClickThroughEvent(&outputXML, creativeId, false)
+				encoder.EncodeToken(tt)
+			case "Creative":
+				isCreative = false
+			case "CompanionAds":
+				if !companionTagFound && wrapperTagFound {
+					encoder.Flush()
+					ti.addCompanionClickThroughEvent(&outputXML, creativeId, true)
+				}
+			}
+		case xml.CharData:
+			tt2 := strings.Trim(string(tt), trimRunes)
+			if len(tt2) != 0 {
+				encoder.Flush()
+				outputXML.WriteString("<![CDATA[")
+				outputXML.WriteString(tt2)
+				outputXML.WriteString("]]>")
+				continue
+			}
+		}
+
+		encoder.EncodeToken(t)
+	}
+
+	encoder.Flush()
+
+	if !inlineWrapperTagFound {
+		// 	// Todo log adapter.<bidder-name>.requests.badserverresponse metrics
+		return vastXML
+	}
+	return outputXML.String()
+}
+
+func (ti *TrackerInjector) addTrackingEvent(outputXML *strings.Builder, creativeId string, addParentTag bool) {
+	if addParentTag {
+		outputXML.WriteString("<TrackingEvents>")
+	}
+	for typ, urls := range ti.events.LinearTrackingEvents {
+		ti.writeTrackingEvent(urls, outputXML, "<Tracking event=\""+string(typ)+"\"><![CDATA[", "]]></Tracking>", creativeId, typ, "tracking")
+	}
+	if addParentTag {
+		outputXML.WriteString("</TrackingEvents>")
+	}
+}
+
+func (ti *TrackerInjector) addClickTrackingEvent(outputXML *strings.Builder, creativeId string, addParentTag bool) {
+	if addParentTag {
+		outputXML.WriteString("<VideoClicks>")
+	}
+	ti.writeTrackingEvent(ti.events.VideoClicks, outputXML, "<ClickTracking><![CDATA[", "]]></ClickTracking>", creativeId, "", "clicktracking")
```

Comment:
I wonder for readability, if various Start/End tags throughout these functions like ClickTracking, /ClickTracking, /Tracking, ![CDATA[", "]], etc. could be extracted to constants.

What do you think?

Reply:
Done